File: examples/05_gripper_opening_control.py
"""Tier 3 -- normalized gripper opening control.

This example demonstrates the formal gripper API: opening `0.0` is fully closed
and opening `1.0` is fully open. Entering the robot context connects, enables,
and calibrates gripper opening before user code commands `open`, `set_opening`,
and `close`.

Usage:
    python examples/05_gripper_opening_control.py --mock \
        --i-understand-this-moves-the-gripper --seconds 0.1

    python examples/05_gripper_opening_control.py --interface can0 \
        --send-id 0x08 --recv-id 0x18 --motor-type DM4310 \
        --opening-direction increasing_position \
        --i-understand-this-moves-the-gripper
"""

from __future__ import annotations

import argparse
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from examples._bringup_common import (
    add_bounded_run_args,
    add_fd_arg,
    add_interface_arg,
    add_single_motor_args,
    print_assumptions,
    resolve_motor_type,
    validate_seconds,
)


def tick_for(robot, seconds: float, deadline_us: int) -> int:
    deadline = time.monotonic() + seconds
    ticks = 0
    while time.monotonic() < deadline:
        robot.tick(deadline_us)
        ticks += 1
        time.sleep(0.001)
    return ticks


def main() -> int:
    parser = argparse.ArgumentParser(description=(__doc__ or "").splitlines()[0])
    parser.add_argument("--mock", action="store_true", help="use MockCanBus (no hardware)")
    parser.add_argument(
        "--opening-direction",
        choices=("increasing_position", "decreasing_position"),
        default="increasing_position",
        help="whether increasing raw motor position opens the gripper",
    )
    parser.add_argument(
        "--default-current",
        type=float,
        default=0.15,
        help="default opening current in per-unit motor current (default: 0.15)",
    )
    parser.add_argument(
        "--current",
        type=float,
        default=None,
        help="optional per-command current override for the midpoint command",
    )
    parser.add_argument(
        "--i-understand-this-moves-the-gripper",
        action="store_true",
        help="required safety acknowledgement",
    )
    add_interface_arg(parser)
    add_fd_arg(parser)
    add_single_motor_args(
        parser,
        default_send_id=0x08,
        default_recv_id=0x18,
        default_motor_type="DM4310",
    )
    add_bounded_run_args(parser, default_seconds=2.0)
    args = parser.parse_args()
    validate_seconds(args, max_seconds=10.0)

    if not args.i_understand_this_moves_the_gripper:
        raise SystemExit("Pass --i-understand-this-moves-the-gripper to run this example.")
    if args.default_current <= 0.0 or args.default_current > 1.0:
        raise SystemExit("--default-current must be > 0.0 and <= 1.0")
    if args.current is not None and (args.current <= 0.0 or args.current > 1.0):
        raise SystemExit("--current must be > 0.0 and <= 1.0")

    motor_type = resolve_motor_type(args.motor_type)
    print_assumptions(
        "05 gripper normalized opening control",
        [
            f"interface          : {args.interface}",
            f"target             : {'MockCanBus' if args.mock else args.interface}",
            f"send_id / recv_id  : 0x{args.send_id:02X} / 0x{args.recv_id:02X}",
            f"motor_type         : {args.motor_type}",
            f"opening_direction  : {args.opening_direction}",
            f"default_current    : {args.default_current}",
            f"duration           : {args.seconds:.2f} s per opening (hard cap 10 s)",
            "lifecycle          : connect -> enable -> automatic opening calibration -> commands",
            "commands           : open() -> set_opening(0.5) -> close()",
            "safety             : clear the gripper jaws before sending commands",
        ],
        fd=args.fd,
    )

    import can_motor_control
    from can_motor_control.damiao import DamiaoCodec

    transport = (
        can_motor_control.MockCanBus.new_fd("vcan_mock")
        if args.mock and args.fd
        else can_motor_control.MockCanBus("vcan_mock")
        if args.mock
        else can_motor_control.SocketCanBus(args.interface, fd=args.fd)
    )
    codec = can_motor_control.MockFeedbackCodec(args.recv_id) if args.mock else DamiaoCodec()
    robot = (
        can_motor_control.RobotBuilder()
        .add_bus("main", transport, codec)
        .add_gripper(
            "grip",
            bus="main",
            motor=can_motor_control.MotorSpec("g", motor_type, args.send_id, args.recv_id),
            opening_direction=args.opening_direction,
            default_current=args.default_current,
        )
        .build()
    )

    with robot:
        gripper = robot["grip"]
        print("commanding open()...")
        gripper.open()
        ticks = tick_for(robot, args.seconds, args.deadline_us)

        print("commanding set_opening(0.5)...")
        gripper.set_opening(0.5, current=args.current)
        ticks += tick_for(robot, args.seconds, args.deadline_us)

        print("commanding close()...")
        gripper.close()
        ticks += tick_for(robot, args.seconds, args.deadline_us)

        motor = gripper.motor
        print(
            f"completed {ticks} ticks; gripper "
            f"pos={motor.position:+.5f} vel={motor.velocity:+.5f} tau={motor.torque:+.5f} "
            f"t_mos={motor.temperature_mos} t_rotor={motor.temperature_rotor}"
        )
    print("done.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/can-motor-control/examples/05_gripper_opening_control.rs
//! Tier 3 Rust bring-up: normalized gripper opening control.
//!
//! Opening `0.0` is fully closed and opening `1.0` is fully open. `Robot::enable`
//! runs opening calibration before `open`, `set_opening`, and `close` are used.

use std::env;
use std::io::{self, Write};
use std::process::ExitCode;
use std::sync::Mutex;
use std::thread;
use std::time::{Duration, Instant};

use can_motor_control::{
    BusCapabilities, CanBus, CanFrame, CodecError, Command, Event, GripperOpeningSpec, Limits,
    MockCanBus, MotorCodec, MotorRef, MotorSpec, MotorTypeId, OpeningDirection, RobotBuilder,
    SocketCanBus,
};
use damiao_codec::{parse_motor_type, DamiaoCodec};

const DEFAULT_INTERFACE: &str = "can0";
const DEFAULT_SEND_ID: u32 = 0x08;
const DEFAULT_RECV_ID: u32 = 0x18;
const DEFAULT_MOTOR_TYPE: &str = "DM4310";
const DEFAULT_SECONDS: f64 = 2.0;
const DEFAULT_DEADLINE_US: u64 = 500;
const DEFAULT_CURRENT: f64 = 0.15;

struct MockFeedbackCodec {
    recv_id: u32,
    position: Mutex<f64>,
}

impl MockFeedbackCodec {
    fn new(recv_id: u32) -> Self {
        Self {
            recv_id,
            position: Mutex::new(0.0),
        }
    }
}

impl MotorCodec for MockFeedbackCodec {
    fn vendor_name(&self) -> &'static str {
        "mock-feedback"
    }

    fn supports(&self, t: MotorTypeId) -> bool {
        matches!(t, MotorTypeId::Damiao(_))
    }

    fn limits(&self, _: MotorTypeId) -> Result<Limits, CodecError> {
        Ok(Limits {
            p_max: 1.0,
            v_max: 1.0,
            t_max: 1.0,
        })
    }

    fn bind_to_bus(&mut self, _: BusCapabilities) {}

    fn encode_enable(&self, m: MotorRef<'_>) -> Result<CanFrame, CodecError> {
        CanFrame::classical(m.send_id, &[0xFC])
            .map_err(|_| CodecError::DecodeFailed { reason: "frame" })
    }

    fn encode_disable(&self, m: MotorRef<'_>) -> Result<CanFrame, CodecError> {
        CanFrame::classical(m.send_id, &[0xFD])
            .map_err(|_| CodecError::DecodeFailed { reason: "frame" })
    }

    fn encode_set_zero(&self, m: MotorRef<'_>) -> Result<CanFrame, CodecError> {
        if let Ok(mut position) = self.position.lock() {
            *position = 0.0;
        }
        CanFrame::classical(m.recv_id, &[0xFE])
            .map_err(|_| CodecError::DecodeFailed { reason: "frame" })
    }

    fn encode_command(&self, m: MotorRef<'_>, command: &Command) -> Result<CanFrame, CodecError> {
        if let Command::PosForce { q, .. } = *command {
            if let Ok(mut position) = self.position.lock() {
                *position = q;
            }
            CanFrame::classical(m.recv_id, &[0x55])
                .map_err(|_| CodecError::DecodeFailed { reason: "frame" })
        } else {
            CanFrame::classical(m.send_id, &[0x55])
                .map_err(|_| CodecError::DecodeFailed { reason: "frame" })
        }
    }

    fn decode(&self, frame: &CanFrame) -> Result<Option<Event>, CodecError> {
        if frame.id != self.recv_id {
            return Ok(None);
        }
        let q = *self.position.lock().map_err(|_| CodecError::DecodeFailed {
            reason: "mock lock",
        })?;
        Ok(Some(Event::State {
            motor_id: self.recv_id,
            q,
            dq: 0.0,
            tau: 0.0,
            t_mos: 30,
            t_rotor: 35,
        }))
    }
}

struct Args {
    mock: bool,
    interface: String,
    fd: bool,
    send_id: u32,
    recv_id: u32,
    motor_type: String,
    seconds: f64,
    deadline_us: u64,
    opening_direction: OpeningDirection,
    default_current: f64,
    current: Option<f64>,
    acknowledged: bool,
}

fn print_help() {
    println!(
        "Tier 3 Rust bring-up: normalized gripper opening control.\n\
\n\
Usage:\n\
  cargo run -p can-motor-control --example 05_gripper_opening_control -- [options]\n\
\n\
Options:\n\
  --mock                                  Use MockCanBus (no hardware)\n\
  --interface <iface>                     SocketCAN interface name (default: can0)\n\
  --fd                                    Open the bus in CAN-FD mode\n\
  --send-id <id>                          CAN id host->motor (default: 0x08)\n\
  --recv-id <id>                          CAN id motor->host (default: 0x18)\n\
  --motor-type <sku>                      Damiao motor model (default: DM4310)\n\
  --opening-direction <increasing_position|decreasing_position>\n\
  --default-current <value>               Default opening current per-unit (default: 0.15)\n\
  --current <value>                       Optional midpoint command current override\n\
  --seconds <seconds>                     Per-opening duration, >0 and <=10\n\
  --deadline-us <us>                      Per tick bus deadline in microseconds\n\
  --i-understand-this-moves-the-gripper   Required safety acknowledgement\n\
  -h, --help                              Print this help text"
    );
}

fn parse_can_id(text: &str) -> Result<u32, String> {
    let trimmed = text.trim();
    if let Some(hex) = trimmed
        .strip_prefix("0x")
        .or_else(|| trimmed.strip_prefix("0X"))
    {
        u32::from_str_radix(hex, 16).map_err(|_| format!("invalid CAN id: {text}"))
    } else {
        trimmed
            .parse::<u32>()
            .map_err(|_| format!("invalid CAN id: {text}"))
    }
}

fn next_value(args: &mut impl Iterator<Item = String>, flag: &str) -> Result<String, String> {
    args.next()
        .ok_or_else(|| format!("{flag} requires a value"))
}

fn parse_f64(args: &mut impl Iterator<Item = String>, flag: &str) -> Result<f64, String> {
    let value = next_value(args, flag)?;
    value
        .parse::<f64>()
        .map_err(|_| format!("invalid {flag} value: {value}"))
}

fn parse_opening_direction(text: &str) -> Result<OpeningDirection, String> {
    match text {
        "increasing_position" => Ok(OpeningDirection::IncreasingPosition),
        "decreasing_position" => Ok(OpeningDirection::DecreasingPosition),
        _ => Err(format!(
            "unknown --opening-direction {text}; expected increasing_position or decreasing_position"
        )),
    }
}

fn parse_args() -> Result<Option<Args>, String> {
    let mut parsed = Args {
        mock: false,
        interface: DEFAULT_INTERFACE.to_string(),
        fd: false,
        send_id: DEFAULT_SEND_ID,
        recv_id: DEFAULT_RECV_ID,
        motor_type: DEFAULT_MOTOR_TYPE.to_string(),
        seconds: DEFAULT_SECONDS,
        deadline_us: DEFAULT_DEADLINE_US,
        opening_direction: OpeningDirection::IncreasingPosition,
        default_current: DEFAULT_CURRENT,
        current: None,
        acknowledged: false,
    };
    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "-h" | "--help" => return Ok(None),
            "--mock" => parsed.mock = true,
            "--fd" => parsed.fd = true,
            "--interface" => parsed.interface = next_value(&mut args, "--interface")?,
            "--send-id" => parsed.send_id = parse_can_id(&next_value(&mut args, "--send-id")?)?,
            "--recv-id" => parsed.recv_id = parse_can_id(&next_value(&mut args, "--recv-id")?)?,
            "--motor-type" => parsed.motor_type = next_value(&mut args, "--motor-type")?,
            "--seconds" => parsed.seconds = parse_f64(&mut args, "--seconds")?,
            "--deadline-us" => {
                let value = next_value(&mut args, "--deadline-us")?;
                parsed.deadline_us = value
                    .parse::<u64>()
                    .map_err(|_| format!("invalid --deadline-us value: {value}"))?;
            }
            "--opening-direction" => {
                parsed.opening_direction =
                    parse_opening_direction(&next_value(&mut args, "--opening-direction")?)?;
            }
            "--default-current" => {
                parsed.default_current = parse_f64(&mut args, "--default-current")?
            }
            "--current" => parsed.current = Some(parse_f64(&mut args, "--current")?),
            "--i-understand-this-moves-the-gripper" => parsed.acknowledged = true,
            _ => return Err(format!("unknown argument: {arg}")),
        }
    }
    if !parsed.acknowledged {
        return Err("pass --i-understand-this-moves-the-gripper to run this example".to_string());
    }
    if parsed.seconds <= 0.0 || parsed.seconds > 10.0 {
        return Err(format!(
            "--seconds must be > 0 and <= 10, got {}",
            parsed.seconds
        ));
    }
    if parsed.default_current <= 0.0 || parsed.default_current > 1.0 {
        return Err("--default-current must be > 0.0 and <= 1.0".to_string());
    }
    if let Some(current) = parsed.current {
        if current <= 0.0 || current > 1.0 {
            return Err("--current must be > 0.0 and <= 1.0".to_string());
        }
    }
    Ok(Some(parsed))
}

fn tick_for(
    robot: &mut can_motor_control::Robot,
    seconds: f64,
    deadline_us: u64,
) -> Result<usize, Box<dyn std::error::Error>> {
    let deadline = Instant::now() + Duration::from_secs_f64(seconds);
    let tick_deadline = Duration::from_micros(deadline_us);
    let mut ticks = 0;
    while Instant::now() < deadline {
        robot.tick(tick_deadline)?;
        ticks += 1;
        thread::sleep(Duration::from_millis(1));
    }
    Ok(ticks)
}

fn run(args: Args) -> Result<(), Box<dyn std::error::Error>> {
    let motor_type = parse_motor_type(&args.motor_type)
        .ok_or_else(|| format!("unknown --motor-type {}", args.motor_type))?;
    println!("=== Rust 05 gripper normalized opening control ===");
    println!("  - interface          : {}", args.interface);
    println!(
        "  - target             : {}",
        if args.mock {
            "MockCanBus"
        } else {
            &args.interface
        }
    );
    println!(
        "  - send_id / recv_id  : 0x{:02X} / 0x{:02X}",
        args.send_id, args.recv_id
    );
    println!("  - motor_type         : {}", args.motor_type);
    println!("  - default_current    : {}", args.default_current);
    println!(
        "  - lifecycle          : connect -> enable -> automatic opening calibration -> commands"
    );
    println!("  - commands           : open() -> set_opening(0.5) -> close()");
    println!("  - safety             : clear the gripper jaws before sending commands");
    let _ = io::stdout().flush();

    let transport: Box<dyn CanBus> = if args.mock && args.fd {
        Box::new(MockCanBus::new_fd("vcan_mock"))
    } else if args.mock {
        Box::new(MockCanBus::new("vcan_mock"))
    } else {
        Box::new(SocketCanBus::open(&args.interface, args.fd)?)
    };
    let codec: Box<dyn MotorCodec> = if args.mock {
        Box::new(MockFeedbackCodec::new(args.recv_id))
    } else {
        Box::new(DamiaoCodec::new())
    };
    let mut robot = RobotBuilder::new()
        .add_bus("main", transport, codec)
        .add_gripper_with_opening(
            "grip",
            "main",
            MotorSpec::new("g", motor_type, args.send_id, args.recv_id),
            GripperOpeningSpec::new(args.opening_direction, Some(args.default_current)),
        )
        .build()?;

    robot.connect()?;
    robot.enable()?;

    {
        let gripper = robot
            .group_mut("grip")
            .and_then(|group| group.as_gripper_mut())
            .ok_or("missing gripper group")?;
        gripper.open(None)?;
    }
    let mut ticks = tick_for(&mut robot, args.seconds, args.deadline_us)?;
    {
        let gripper = robot
            .group_mut("grip")
            .and_then(|group| group.as_gripper_mut())
            .ok_or("missing gripper group")?;
        gripper.set_opening(0.5, args.current)?;
    }
    ticks += tick_for(&mut robot, args.seconds, args.deadline_us)?;
    {
        let gripper = robot
            .group_mut("grip")
            .and_then(|group| group.as_gripper_mut())
            .ok_or("missing gripper group")?;
        gripper.close(None)?;
    }
    ticks += tick_for(&mut robot, args.seconds, args.deadline_us)?;
    robot.disable()?;

    println!("completed {ticks} ticks; done.");
    Ok(())
}

fn main() -> ExitCode {
    let args = match parse_args() {
        Ok(Some(args)) => args,
        Ok(None) => {
            print_help();
            return ExitCode::SUCCESS;
        }
        Err(message) => {
            eprintln!("ERROR: {message}");
            eprintln!("Try --help for usage.");
            return ExitCode::from(2);
        }
    };
    match run(args) {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("ERROR: {error}");
            ExitCode::FAILURE
        }
    }
}
